src/components/application_manager/include/application_manager/help_prompt_manager_impl.h

```diff
@@ -0,0 +1,149 @@
+ * by the user when the application does not provide any or such information.
+ */
+class HelpPromptManagerImpl : public HelpPromptManager {
+ public:
+  /**
+   * @brief Continer for buffering VR help commands
+  */
+  typedef std::map<uint32_t, smart_objects::SmartObjectSPtr> VRCommandsMap;
+  /**
+   * @brief Class constructor
+   * @param app owner for this class
+   */
+  HelpPromptManagerImpl(Application& app, ApplicationManager& app_manager);
+  /**
+   * @brief Class destructor
+   */
+  ~HelpPromptManagerImpl();
+  /**
+   * @brief The type sends info to HMI
+   */
+  enum class SendingType { kNoneSend, kSendHelpPrompt, kSendVRHelp, kSendBoth };
+
+  /**
+   * @brief Adds command to constructed values, and send SetGlobalProperties if
+   * required
+   */
+  void OnVrCommandAdded(const uint32_t cmd_id,
+                        const smart_objects::SmartObject& command) OVERRIDE;
+  /**
+   * @brief Removes command from constructed values, and send
+   * SetGlobalProperties if required
+   */
+  void OnVrCommandDeleted(const uint32_t cmd_id) OVERRIDE;
+  /**
+     * @brief Stop constructing vrHelp and/or helpPrompt if they are present in
+     * message
+     * @param msg containing GlobalProperties
+     * @param is_response determines is the request or response for the for
+     * the SetGlobalPropertiesRequest
+     */
+  void OnSetGlobalPropertiesReceived(const smart_objects::SmartObject& msg,
+                                     const bool is_response) OVERRIDE;
+  /**
+    * @brief Starts mechanism for handling "help" requests
+    * @param is_restore determines the need to restore commands
+    * after resuming
+    */
+  void OnAppActivated(const bool is_restore) OVERRIDE;
+  /**
+    * @brief Stop mechanism for handling "help" requests
+    */
+  void OnAppUnregistered() OVERRIDE;
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(HelpPromptManagerImpl);
+  /**
+   * @brief Send TTS request to HMI
+   */
+  void SendTTSRequest();
+  /**
+   * @brief Send UI request to HMI
+   */
+  void SendUIRequest();
+  /**
+   * @brief Send TTS and UI requests to HMI
+   */
+  void SendBothRequests();
+  /**
+   * @brief Send TTS or UI or both Requests
+   */
+  void SendRequests();
+  /**
+   * @brief Construct the helpPrompt parameter
+   */
+  void CreatePromptMsg(smart_objects::SmartObject& out_msg_params);
+  /**
+   * @brief Construct the vrHelp parameter
+   */
+  void CreateVRMsg(smart_objects::SmartObject& out_msg_params);
+  /**
+   * @brief Getting request type to send HMI
+   * @param msg containing request or response for the
+   * SetGlobalPropertiesRequest
+   * @param is_response determines is the request or response for the for
+   * the SetGlobalPropertiesRequest
+   * @return SendingType
+   */
+  SendingType GetSendingType(const smart_objects::SmartObject& msg,
+                             const bool is_response);
+
+  Application& app_;
+  ApplicationManager& app_manager_;
+  VRCommandsMap vr_commands_;
```

Comment:
Should we include a lock for this map?

Reply:
@jacobkeeler added just for case when some thread is adding/deleting command and OnAppActivated come from another thread. See https://github.com/smartdevicelink/sdl_core/pull/2199/commits/56b01ce0d32d1a4059125a350ad108d35c529c9f